### PR TITLE
Only clear valid ABI files

### PR DESCRIPTION
**Pull request by ItsNickBarry**
Continuation of #21.

### CODE REVIEW

**Comment by ItsNickBarry** [comment]
@will-holley I implemented option 1 from our discussion in PR #21.  Looks okay?

**Comment by ItsNickBarry** [comment]
New import allowed removal of custom `prettify` function as well.

File: package.json [hunk 1]
```diff
@@ -23,6 +23,10 @@
   "peerDependencies": {
     "hardhat": "^2.0.0"
   },
+  "dependencies": {
+    "@ethersproject/abi": "^5.5.0",
```
Comment (will-holley) [inline package.json:27]: These should be dev dependencies.

File: index.js [hunk 2]
```diff
@@ -72,22 +47,29 @@ task(TASK_COMPILE, async function (args, hre, runSuper) {
...
   if (config.clear) {
-    if (fs.existsSync(outputDirectory)) {
-      fs.rmSync(outputDirectory, { recursive: true });
+    const files = readdirRecursive(outputDirectory).filter(f => path.extname(f) === '.json');
+
+    for (let file of files) {
+      try {
+        const filepath = path.resolve(outputDirectory, file);
+        const contents = fs.readFileSync(filepath).toString();
+        new Interface(contents);
```
Comment (will-holley) [inline index.js:65]: This validates that the JSON file is an ABI (by throwing if it is not), correct?

**Review by will-holley — APPROVED** [review]
1 requested change and 1 question – otherwise, LGTM.

File: package.json (hunk 1 above)
Comment (ItsNickBarry) [reply package.json:27]: Dev dependencies are _only_ installed in a local clone of the repository when `yarn install` is run.  These packages need to be included in end users' projects.  Whether `hardhat-abi-exporter` itself is installed as a dev dependency or normal dependency, only its normal dependencies will be installed along with it.

File: index.js (hunk 2 above)
Comment (ItsNickBarry) [reply index.js:65]: Yes.  I haven't been able to find a validator that returns true/false, so we just rely on the `Interface` initializer to fail if the passed data is not a valid ABI.

**Comment by ItsNickBarry** [comment]
@will-holley released in version `2.4.0`.  Note the addition of the `runOnCompile` option (#23), which is now required to enable behavior that was previously the default.

File: index.js (hunk 2 above)
Comment (sambacha) [reply index.js:65]: > Yes. I haven't been able to find a validator that returns true/false, so we just rely on the `Interface` initializer to fail if the passed data is not a valid ABI.

```js
const { Interface } = require('@ethersproject/abi');
const { Signer } = require('@ethersproject/abstract-signer');
const { isAddress } = require('@ethersproject/address');
const { Contract } = require('@ethersproject/contracts');
const { LedgerSigner } = require('@ethersproject/hardware-wallets');
const { isValidName } = require('@ethersproject/hash');
const { isValidMnemonic, defaultPath } = require('@ethersproject/hdnode');
const {
  getDefaultProvider,
  JsonRpcProvider,
} = require('@ethersproject/providers');
const { Wallet } = require('@ethersproject/wallet');
const getFunctionArgs = require('./utils/getParams.js');
const formatResult = require('./utils/formatResult.js');
(async () => {
  const argv = yargs.option('address', { string: true }).argv;
  prompts.override(argv);

  const { artifact } = await prompts({
    type: 'text',
    name: 'artifact',
    message: 'Path to artifact',
    validate: fs.existsSync,
    format: (path) => JSON.parse(fs.readFileSync(path)),
    initial: 'build/contracts/YourContractABI.json',
  });
  if (artifact == undefined) {
    throw 'Undefined Artifact, Terminating';
  }
  if (!artifact.abi) {
    throw 'Invalid artifact - not of type ABI';
  }
```

This is what I use and its fairly accurate. Note the code is truncated only for the relevant part for you, you can find the rest of it located here: https://github.com/manifoldfinance/libcaller/tree/master/packages/gencall-cli

Comment (ItsNickBarry) [reply index.js:65]: @sambacha Seems like that code validates whether an artifact contains an ABI, but what we need here is to validate that a JSON file is a valid ABI after it has already been extracted from the artifact.  I'm fine with the `Interface` approach used in this PR - if the @ethersproject/abi packagesis updated to include an internal validator in the future, I'll update this package to make use of it.

Comment (sambacha) [reply index.js:65]: mmm now that I think about it, would it actually be possible to offer Truffles ABI format as an option. for output format? https://github.com/trufflesuite/truffle/tree/develop/packages/abi-utils 

Cheers,

Sam

Comment (ItsNickBarry) [reply index.js:65]: That's a good idea, I'll look into it.

Comment (ItsNickBarry) [reply index.js:65]: Seems like Hardhat ABIs are already normalized.

**Pull request merged**